File: day7/d7.js
const fs = require('fs');

function evaluateExpression(numbers, operators) {
    let result = numbers[0];
    for (let i = 0; i < operators.length; i++) {
        if (operators[i] === '+') {
            result += numbers[i + 1];
        } else if (operators[i] === '*') {
            result *= numbers[i + 1];
        }
    }
    return result;
}

function canBeTrue(testValue, numbers) {
    if (numbers.length === 1) {
        return numbers[0] === testValue;
    }

    const operators = ['+', '*'];
    const combinations = getCombinations(operators, numbers.length - 1);

    for (const ops of combinations) {
        if (evaluateExpression(numbers, ops) === testValue) {
            return true;
        }
    }
    return false;
}

function getCombinations(operators, length) {
    if (length === 1) {
        return operators.map(op => [op]);
    }

    const combinations = [];
    const smallerCombinations = getCombinations(operators, length - 1);

    for (const op of operators) {
        for (const smallerCombination of smallerCombinations) {
            combinations.push([op, ...smallerCombination]);
        }
    }

    return combinations;
}

function main() {
    let totalCalibrationResult = 0;
    const data = fs.readFileSync('d7Input.txt', 'utf8');
    const lines = data.split('\n');

    for (const line of lines) {
        if (line.trim() === '') continue;

        const [testValueStr, numbersStr] = line.split(':');
        const testValue = parseInt(testValueStr.trim());
        const numbers = numbersStr.trim().split(' ').map(Number);

        if (canBeTrue(testValue, numbers)) {
            totalCalibrationResult += testValue;
        }
    }

    console.log(totalCalibrationResult);
}

main();

File: day7/d7.py
import itertools

def evaluate_expression(numbers, operators):
    result = numbers[0]
    for i in range(len(operators)):
        if operators[i] == '+':
            result += numbers[i + 1]
        elif operators[i] == '*':
            result *= numbers[i + 1]
    return result

def can_be_true(test_value, numbers):
    if len(numbers) == 1:
        return numbers[0] == test_value
    
    operators = ['+', '*']
    for ops in itertools.product(operators, repeat=len(numbers) - 1):
        if evaluate_expression(numbers, ops) == test_value:
            return True
    return False

def main():
    total_calibration_result = 0
    
    with open('d7Input.txt', 'r') as file:
        for line in file:
            test_value_str, numbers_str = line.split(':')
            test_value = int(test_value_str.strip())
            numbers = list(map(int, numbers_str.strip().split()))
            
            if can_be_true(test_value, numbers):
                total_calibration_result += test_value
    
    print(total_calibration_result)

if __name__ == "__main__":
    main()